rfcs/proposed/parallel_block_for_task_arena/README.md

```diff
@@ -0,0 +1,178 @@
+  * What if we just indicated the end of the "Parallel Block"?
+
+The extended state machine aims to answer these questions.
+* The first call to the “Start of PB” will transition into the “Parallel Block” state.
+* The last call to the “End of PB” will transition back to the “Delayed leave” state
+  or into the "One-time Fast leave" if it is indicated that threads should leave sooner.
+* Concurrent or nested calls to the “Start of PB” or the “End of PB”
+  increment/decrement reference counter.
+
+<img src="parallel_block_state_final.png" width=800>
+
+Let's consider the semantics that an API for explicit parallel blocks can provide:
+* Start of a parallel block:  * Indicates the point from which the scheduler can use a hint and stick threads to the arena.
+  * Serve as a warm-up hint to the scheduler, making some worker threads immediately available
+    at the start of the real computation.
+* "Parallel block" itself:
+  * Scheduler can implement different policies to retain threads in the arena.
+* End of a parallel block:
+  * Indicates the point from which the scheduler can drop a hint and unstick threads from the arena.
+  * Indicates that arena should enter the “One-time Fast leave” thus workers can leave sooner.
+
+
+Start of a parallel block:<br>
+The warm-up hint should have similar guarantees as `task_arena::enqueue` from a signal standpoint.
+Users should expect the scheduler will do its best to make some threads available in the arena.
+
+"Parallel block" itself:<br>
+The semantic for retaining threads is a hint to the scheduler;
+thus, no real guarantee is provided. The scheduler can ignore the hint and
+move threads to another arena or to sleep if conditions are met.
+
+End of a parallel block:<br>
+Can indicate that arena should enter the “One-time Fast leave” thus workers can leave sooner.
+However, if work was submitted immediately after the end of the parallel block,
+the default arena "workers leave" state will be restored.
+If the default "workers leave" state was the "Fast leave" the result is NOP.
+
+```cpp
+class task_arena {
+    enum class workers_leave {
+        fast,
+        delayed
+    };
+
+    task_arena(const constraints& constraints_, unsigned reserved_for_masters = 1,
+               priority a_priority = priority::normal,
+               workers_leave a_workers_leave = workers_leave::delayed);
+
+    void parallel_block_start();
+    void parallel_block_end(bool set_one_time_fast_leave = false);
+    scoped_parallel_block_type scoped_parallel_block(bool set_one_time_fast_leave = false);
+};
+
+namespace this_task_arena {
+    void parallel_block_start();
+    void parallel_block_end(bool set_one_time_fast_leave = false);
+    scoped_parallel_block_type scoped_parallel_block(bool set_one_time_fast_leave = false);
+}
+```
+
+By the contract, users should call `parallel block start` for each
+subsequent call to `parallel block_end`.<br>
```

Comment:
I would say the other way around: users should indicate the end of block for each previous start of block.

Reply:
Done.